LempelZivWelch

RomanLevashev opened the pull request.

RomanLevashev committed:
Added LZW and ByteTrie classes for compression and byte sequence storage.

RomanLevashev committed:
Comments Add

RomanLevashev committed:
Tests for ByteTrie and LZW added

RomanLevashev committed:
Added compression ratio return value to CompressFile method
Added command-line argument parsing for LZW compression program

yurii-litvinov commented on LempelZivWelch/ByteTrie/ByteTrie.csproj:
```diff
@@ -0,0 +1,16 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
```
Актуален .NET 9

yurii-litvinov replied:
Не поправлено

yurii-litvinov commented on LempelZivWelch/ByteTrie/ByteTrie.csproj:
```diff
@@ -0,0 +1,16 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="StyleCop.Analyzers" Version="1.1.118">
```
Используйте StyleCop 1.2-beta, 1.1 не знает синтаксиса новых версий C#

yurii-litvinov commented on LempelZivWelch/ByteTrie/Node.cs:
```diff
@@ -0,0 +1,36 @@
+﻿namespace ByteTrie
```
Используйте file-scoped namespaces

yurii-litvinov commented on LempelZivWelch/ByteTrie/Node.cs:
```diff
@@ -0,0 +1,36 @@
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
```
Надо убирать лишние using-и, тем более что все нормальные среды разработки могут автоматически убирать using-и во всём проекте сразу.

yurii-litvinov commented on LempelZivWelch/ByteTrie/Node.cs:
```diff
@@ -0,0 +1,36 @@
+    /// Represents a node in the trie data structure.
+    /// </summary>
+    /// <param name="value">The byte value associated with the node.</param>
+    public class Node(byte value)
```
Сделайте его хотя бы internal. Внешнему миру вовсе не обязательно про него знать

yurii-litvinov commented on LempelZivWelch/ByteTrie/Trie.cs:
```diff
@@ -0,0 +1,151 @@
+    using static System.Runtime.InteropServices.JavaScript.JSType;
+
+    /// <summary>
+    /// Представляет собой префиксное дерево (бор) для хранения строк.
```
Один комментарий по-русски и остальные по-английски — выглядит странно

yurii-litvinov commented on LempelZivWelch/ByteTrie/Trie.cs:
```diff
@@ -0,0 +1,151 @@
+        /// <summary>
+        /// Gets the root node of the trie.
+        /// </summary>
+        public Node Root { get;  } = new Node(0);
```
Два пробела вместо одного перед `}` :)

yurii-litvinov commented on LempelZivWelch/ByteTrie/Trie.cs:
```diff
@@ -0,0 +1,151 @@
+        public (Node? terminal, bool isSuccess) Add(byte element)
+        {
+            return this.Add([element]);
+        }
```
```suggestion
        public (Node? terminal, bool isSuccess) Add(byte element)
            => this.Add([element]);
```

yurii-litvinov commented on LempelZivWelch/ByteTrie/Trie.cs:
```diff
@@ -0,0 +1,151 @@
+        {
+            if (element == null || element.Length == 0)
+            {
+                throw new ArgumentException("The element cannot be null or empty.", nameof(element));
```
Лучше отдельно ArgumentNullException.ThrowIfNull(element) и отдельно проверка на element.Length == 0, чтобы мы по сообщению об ошибке точно знали, что не так.

yurii-litvinov commented on LempelZivWelch/ByteTrie/Trie.cs:
```diff
@@ -0,0 +1,151 @@
+        {
+            var (endPrefixNode, nextPosition) = this.FindLongestPrefix(element);
+
+            return nextPosition == element.Length && endPrefixNode.IsTerminal == true;
```
```suggestion
            return nextPosition == element.Length && endPrefixNode.IsTerminal;
```

yurii-litvinov commented on LempelZivWelch/ByteTrie/Trie.cs:
```diff
@@ -0,0 +1,151 @@
+        public bool Contains(byte element)
+        {
+            return this.Contains([element]);
+        }
```
Тут тоже через `=>` лучше

yurii-litvinov commented on LempelZivWelch/ByteTrie/Trie.cs:
```diff
@@ -0,0 +1,151 @@
+
+        private Node CreateSuffix(byte[] element)
+        {
+            Node source = new Node(element[0]);
```
```suggestion
            Node source = new(element[0]);
```
Не пишите один тип дважды.

yurii-litvinov commented on LempelZivWelch/ByteTrie/Trie.cs:
```diff
@@ -0,0 +1,151 @@
+        {
+            int currentPosition = 0;
+            Node currentNode = this.Root;
+            Stack<Node> stack = new Stack<Node>();
```
А он вроде не используется нигде

yurii-litvinov commented on LempelZivWelch/LempelZivWelch/LZW.cs:
```diff
@@ -0,0 +1,370 @@
+    using System.Runtime.InteropServices;
+    using System.Text;
+    using System.Text.Json.Serialization.Metadata;
+    using System.Threading.Tasks;
```
Это даже интересно, откуда взялось — не все из этих пространств имён есть в шаблоне, их надо было вручную подключать :)

yurii-litvinov commented on LempelZivWelch/LempelZivWelch/LZW.cs:
```diff
@@ -0,0 +1,370 @@
+
+                    if (!isSuccess)
+                    {
+                        throw new Exception("Encode error");
```
Прост Exception кидать плохо, объявили бы своё какое-нибудь

yurii-litvinov commented on LempelZivWelch/LempelZivWelch/LZW.cs:
```diff
@@ -0,0 +1,370 @@
+                    }
+
+                    uint index = terminal!.Index;
+                    bool[] binaryInterptitation = GetBinaryInterpritationArray(index, currentCodeSize);
```
Interptitation — опечатка, скорее всего.

yurii-litvinov commented on LempelZivWelch/LempelZivWelch/LZW.cs:
```diff
@@ -0,0 +1,370 @@
+        private static void DecodeBytes(FileStream inputFileStream, FileStream outputFileStream, int bufferSize)
+        {
+            int currentByteIndex = 0;
+            Dictionary<uint, byte[]> sequencesDict = [];
```
Не сокращайте имена

yurii-litvinov commented on LempelZivWelch/LempelZivWelch/LZW.cs:
```diff
@@ -0,0 +1,370 @@
+            return isRefill;
+        }
+
+        private static bool FillBuffer(FileStream inputFileStream, ref int bytesRead, byte[] inputBuffer)
```
Архитектурно стоило работу с буфером и битами вынести в отдельный класс, LZW и так сложный. Так можно было бы отдельно тестировать битовые операции

yurii-litvinov commented on LempelZivWelch/LempelZivWelch/LZW.cs:
```diff
@@ -0,0 +1,370 @@
+            }
+
+            return binaryInterpritation;
+        }
```
Это какой-то костыль, конвертить число в строку и потом вычитывать двоичные цифры как символы. Тем более что в стандартной библиотеке наверняка что-то нормальное для этого есть

yurii-litvinov commented on LempelZivWelch/LempelZivWelch/Program.cs:
```diff
@@ -0,0 +1,28 @@
+switch (operation)
+{
+    case "-c":
+
```
Лишняя пустая строчка

yurii-litvinov commented on LempelZivWelch/UnitTests/LZWTests.cs:
```diff
@@ -0,0 +1,40 @@
+            LZW.DecompressFile(originalFilePath + ".zipped");
+            byte[] decompressedFileBytes = File.ReadAllBytes(originalFilePath);
+
+            Assert.IsTrue(originalFileBytes.Length == decompressedFileBytes.Length);
```
Assert.AreEqual или, что лучше, Assert.That(..., Is.EqualTo(...)). Иначе сообщение об ошибке в случае, если тест не пройдёт, будет невнятным

RomanLevashev committed:
refactor: move buffer handling to separate project and adjust LZW implementation

- Extracted buffer-related logic into a dedicated project for better separation of concerns
- Adapted LZW algorithm to work with the new buffer handling implementation
- Minor style guide compliance adjustments across affected files

RomanLevashev committed:
Add DataHandler reference in LZW project

RomanLevashev committed:
stylecop.json warning removed
Small improvements

yurii-litvinov commented on LempelZivWelch/DataHandler/DecodeDataHandler.cs:
```diff
@@ -0,0 +1,102 @@
+    public override bool IsFileReadComplete
+    {
+        get => base.IsFileReadComplete && this.bitBuffer.Count < this.ChunkSize;
+    }
```
```suggestion
    public override bool IsFileReadComplete => base.IsFileReadComplete && this.bitBuffer.Count < this.ChunkSize;
```
То же самое, но короче

yurii-litvinov commented on LempelZivWelch/DataHandler/EncodeDataHandler.cs:
```diff
@@ -0,0 +1,143 @@
+        {
+            throw new InvalidOperationException(
+            $"Number {num} requires {numBitsNeeded} bits but current chunk size is only {this.ChunkSize} bits. " +
+            "The number is too large for the current encoding settings.");
```
```suggestion
            throw new InvalidOperationException(
                $"Number {num} requires {numBitsNeeded} bits but current chunk size is only {this.ChunkSize} bits. " +
                "The number is too large for the current encoding settings.");
```

yurii-litvinov commented on LempelZivWelch/DataHandler/EncodeDataHandler.cs:
```diff
@@ -0,0 +1,143 @@
+                this.outputBuffer.Clear();
+            }
+
+            bool[] bitArray = new bool[8];
```
```suggestion
            var bitArray = new bool[8];
```

yurii-litvinov commented on LempelZivWelch/LempelZivWelch/LZW.cs:
```diff
@@ -0,0 +1,165 @@
+    /// <returns>
+    /// The compression ratio, which is the ratio of the original file size to the compressed file size.
+    /// </returns>
+    public static long CompressFile(string inputPath)
```
Ожидать, что это число будет целым, не очень правильно

yurii-litvinov reviewed (changes requested):
Декомпрессия не работает — сжал .pdf, разжал, не открывается (и по размеру вдвое больше, чем исходный). Ещё тестов не очень много, но окей.

RomanLevashev committed:
Merge branch 'main' into Lempel-Ziv-Welch

RomanLevashev committed:
.NET 9, workflows

RomanLevashev committed:
Improvements from comments
Add test with pdf file

yurii-litvinov commented on LempelZivWelch/DataHandler/EncodeDataHandler.cs:
```diff
@@ -0,0 +1,133 @@
+    /// <summary>
+    /// Moves the input stream position back by one byte, allowing the last read byte to be read again.
+    /// </summary>
+    public new void ReturnToPreviousByte() => base.ReturnToPreviousByte();
```
Это что-то очень странное, перевводится метод только для того, чтобы делегировать вызов методу базового класса. Можно было бы удалить эту строчку и было бы то же самое.

yurii-litvinov reviewed (approved):
👍
